weasyl/searchtag.py

```diff
@@ -176,152 +247,457 @@ def is_tag_restriction_pattern_valid(text):
     return False
 
 
-def associate(userid, tags, submitid=None, charid=None, journalid=None, preferred_tags_userid=None, optout_tags_userid=None):
+def _update_submission_tags(tx, submitid):
+    tx.execute(
+        "WITH t AS ("
+        " SELECT coalesce(array_agg(tagid), ARRAY[]::integer[]) AS tags FROM searchmapsubmit WHERE targetid = %(submission)s)"
+        " INSERT INTO submission_tags (submitid, tags) SELECT %(submission)s, tags FROM t"
+        " ON CONFLICT (submitid) DO UPDATE SET tags = (SELECT tags FROM t)",
+        submission=submitid,
+    )
+
+
+def _target_unpack(type_code: bytes, id: int) -> TaggableTarget:
+    t = _TARGET_TYPES_BY_CODE.get(type_code)
+
+    if t is None:
+        raise WeasylError("Unexpected")
+
+    return t(id)
+
+
+def _get_ownerid(target: TaggableTarget) -> int | None:
+    return d.get_ownerid(**{target.arg_name: target.id})
+
+
+def _set_suggested_tags(
+    *,
+    userid: int,
+    target: TaggableTarget,
+    ownerid: int,
+    tag_names: set[str],
+) -> list[str]:
+    tagging_privileges_revoked = "g" in d.get_config(userid)
+    if tagging_privileges_revoked:
+        raise WeasylError("InsufficientPermissions")
+
+    if ignoreuser.check(ownerid, userid):
+        raise WeasylError("contentOwnerIgnoredYou")
+
+    # Track which tags fail to be added or removed to later notify the user
+    is_restricted_tag = _get_restricted_tag_matcher({
+        *query_user_restricted_tags(ownerid),
+        *query_global_restricted_tags(),
+    })
+    tags_not_added = set(filter(is_restricted_tag, tag_names))
+    apply_tags = tag_names - tags_not_added
+    apply_tagids = [t.tagid for t in add_and_get_searchtags(apply_tags)]
+
+    def transaction(tx):
+        tx.execute(
+            "WITH"
+            f" removed AS (DELETE FROM {target.table}"
+            " WHERE targetid = %(target)s"
+            " AND settings !~ 'a'"
+            " AND added_by = %(user)s"
+            " AND tagid != ALL (%(tags)s) RETURNING tagid),"
+            f" added AS (INSERT INTO {target.table} (targetid, tagid, added_by)"
```

Comment:
I _think_ the issue is around here, but this looks to allow users to suggest a tag, and given that the `TaggableTarget.table` is updated (so `searchmapsubmit`, for example), this means that a tag that has been suggested--but not approved or rejected--can be searched for. (Unsure if this is intended behavior, or not.)

Consider the following sequence of actions:
- User A has a submission that they created, SubA.
- User B suggests the tag `FooBarBazQuz` to apply to SubA.
- Searching the website via the searchbar for `FooBarBazQuz` will return SubA as a result, despite not being approved or rejected.

For example, consider the following image; we're searching for the tag `qiakwidhchwiodfhdcf` which was suggested as a tag, yet viewing the submission itself shows that the tag does not exist.
![image](https://github.com/Weasyl/weasyl/assets/19718911/954b8f5e-cd76-4677-b11f-9ab3a8163a7b)